pkg/dequeuer/batch_handler.go

```diff
@@ -0,0 +1,216 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dequeuer
+
+import (
+	"bytes"
+	"net/http"
+	"time"
+
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	awslib "github.com/cortexlabs/cortex/pkg/lib/aws"
+	"github.com/cortexlabs/cortex/pkg/lib/errors"
+	"github.com/xtgo/uuid"
+	"go.uber.org/zap"
+)
+
+const _jobCompleteMessageRenewal = 10 * time.Second
+
+type BatchMessageHandler struct {
+	config                    BatchMessageHandlerConfig
+	jobCompleteMessageRenewal time.Duration
+	tags                      []string
+	aws                       *awslib.Client
+	metrics                   statsd.ClientInterface
+	log                       *zap.SugaredLogger
+}
+
+type BatchMessageHandlerConfig struct {
+	APIName   string
+	JobID     string
+	QueueURL  string
+	Region    string
+	TargetURL string
+}
+
+func NewBatchMessageHandler(config BatchMessageHandlerConfig, awsClient *awslib.Client, statsdClient statsd.ClientInterface, log *zap.SugaredLogger) *BatchMessageHandler {
+	tags := []string{
+		"api_name:" + config.APIName,
+		"job_id" + config.JobID,
+	}
+
+	return &BatchMessageHandler{
+		config:                    config,
+		jobCompleteMessageRenewal: _jobCompleteMessageRenewal,
+		tags:                      tags,
+		aws:                       awsClient,
+		metrics:                   statsdClient,
+		log:                       log,
+	}
+}
+
+func (h *BatchMessageHandler) Handle(message *sqs.Message) error {
+	if isOnJobCompleteMessage(message) {
+		err := h.onJobComplete(message)
+		if err != nil {
+			// TODO
+			h.log.Fatal(err)
+		}
+		return nil
+	}
+	err := h.handleBatch(message)
+	if err != nil {
+		// TODO
+		h.log.Errorw("failed processing batch", "id", *message.MessageId, "error", err)
+		err = h.handleFailure(message)
+		if err != nil {
+			// TODO
+			h.log.Fatal(err)
+		}
+	}
+	return nil
+}
+
+func (h *BatchMessageHandler) recordSuccess() error {
+	err := h.metrics.Incr("cortex_batch_succeeded", h.tags, 1.0)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+func (h *BatchMessageHandler) recordFailure() error {
+	err := h.metrics.Incr("cortex_batch_failed", h.tags, 1.0)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+func (h *BatchMessageHandler) recordTimePerBatch(elapsedTime time.Duration) error {
+	err := h.metrics.Histogram("cortex_time_per_batch", elapsedTime.Seconds(), h.tags, 1.0)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+func (h *BatchMessageHandler) submitRequest(messageBody string) error {
+	response, err := http.Post(h.config.TargetURL, "application/json", bytes.NewBuffer([]byte(messageBody)))
```

Comment:
I think it would be good to also set a `"X-Cortex-BatchID"` header. That way, the batch ID can also be retrieved from the request without having to read the `cortex/spec/job.json` file (assuming the user wants the batch ID in the first place).

Reply:
this is done already (I just didn't have pushed when you commented :) )